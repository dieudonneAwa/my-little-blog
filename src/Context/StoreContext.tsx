import React, { useReducer, createContext, useMemo } from 'react';
import { Action } from '../Theme/types';
import useActions, { Actions } from './Actions';
import reducers, { initialStore, InputStoreState } from './Reducers';

export type Props = {
  children: React.ReactNode;
};

export type IStore = {
  actions: Actions;
  state: InputStoreState;
};

const Context = createContext({
  actions: {},
  state: initialStore,
} as IStore);

// The store provider
const Provider = ({ children }: Props) => {
  // Get state and dispatch from Reacts new API useReducer.
  const [reducersState, dispatch] = useReducer(reducers, initialStore);
  const state = useMemo(() => reducersState, [reducersState]);

  // Create an object of all our actions,
  // handling special cases where a simple dispatch is too primitive
  const actions = useActions(dispatch as (arg0: Action) => Action);
  const stateActions = useMemo(() => actions, [actions]);

  // Render state and special case actions
  return (
    <Context.Provider
      // @ts-ignore
      value={{ state, actions: stateActions }}
    >
      {children}
    </Context.Provider>
  );
};

export { Context, Provider };
