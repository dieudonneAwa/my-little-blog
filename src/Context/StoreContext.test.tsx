/* eslint-disable import/no-extraneous-dependencies */
import React from 'react';
import { shallow, ShallowWrapper } from 'enzyme';
import { Props, Provider } from './StoreContext';

const props: Props = {
  children: () => null,
};

describe('Context Provider', (): void => {
  let wrapper: ShallowWrapper<void>;
  beforeEach((): void => {
    wrapper = shallow(<Provider {...props} />);
  });
  it('should render Context Provider component', (): void => {
    expect(wrapper).toMatchSnapshot();
  });
});
