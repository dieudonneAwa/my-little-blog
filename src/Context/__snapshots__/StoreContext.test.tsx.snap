// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Context Provider should render Context Provider component 1`] = `
<ContextProvider
  value={
    Object {
      "actions": Object {
        "fetchArticles": [Function],
        "fetchComments": [Function],
        "fetchUsers": [Function],
        "setArticle": [Function],
      },
      "state": Object {
        "articles": Object {
          "article": Object {},
          "articles": Array [],
          "comments": Array [],
          "loading": false,
          "loadingComments": false,
        },
        "users": Object {
          "error": "",
          "loading": false,
          "user": Object {},
          "users": Array [],
        },
      },
    }
  }
>
  <Component />
</ContextProvider>
`;
